### PR TITLE
Run pytest with different Python versions in GitHub actions

### PR DESCRIPTION
**Pull request by LucyJimenez**
xref: https://github.com/airspeed-velocity/asv/issues/1036
On this PR:
- Add matrix python versions
- Separate test and flake8 on different builds.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,3 +1,5 @@
+
+
```
Comment (datapythonista) [inline .github/workflows/ci.yml:2]: Better to remove those.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -6,6 +8,10 @@ jobs:
   build:
     runs-on: ubuntu-latest
 
+    strategy:
+      matrix:
+        python-version: ["3.7", "3.8", "3.9", "3.10"]
```
Comment (datapythonista) [inline .github/workflows/ci.yml:13]: Just the first and the last versions should be enough. Too many builds add noise and consume resources. Or we may want to add something else to the matrix (like CPython/PyPy), and then the matrix will explode (like 8 builds just for that change.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -19,4 +25,4 @@ jobs:
         run: flake8 .
```
Comment (datapythonista) [inline .github/workflows/ci.yml:25]: I think it's probably better to have flake8 and pytest in separate builds. Doesn't make so much sense to test flake8 in different Python versions.

**Review by datapythonista — COMMENTED** [review]
Thanks for working on this. Looks good, added some minor things.

File: .github/workflows/ci.yml (hunk 3 above)
Comment (LucyJimenez) [reply .github/workflows/ci.yml]: Thank you for the recommendations, I’m going to implement it.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -3,20 +3,25 @@ name: CI
 on: [push, pull_request]
 
 jobs:
-  build:
+  test:
+    name: test w/ Python
```
Comment (datapythonista) [inline .github/workflows/ci.yml:7]: Just tests should be enough. To show that it's not pypy we should say CPython, but pypy is likely to be in the matrix, so better to just name this tests

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -3,20 +3,25 @@ name: CI
...
     runs-on: ubuntu-latest
-
+    strategy:
+      matrix:
+        python-version: ["3.7", "3.10"]
```
Comment (datapythonista) [inline .github/workflows/ci.yml:11]: Did you check that tests are being run with these Python versions?

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -3,20 +3,25 @@ name: CI
...
-      
-      - name: flake8
-        run: flake8 .
-      
+        run: pip install -r requirements-dev.txt
       - name: Run tests
         run: pytest
+  
+  flake8-formatting:
+    name: code autoformatting with flake8
```
Comment (datapythonista) [inline .github/workflows/ci.yml:20]: This is misleading. We are not formatting or autoformatting anything. Linting would be the generic name for checking code styles, flake8 is the tool fot checking them.

**Review by datapythonista — COMMENTED** [review]
Thanks @LucyJimenez, added some comments.

File: .github/workflows/ci.yml (hunk 5 above)
Comment (LucyJimenez) [reply .github/workflows/ci.yml:11]: Yes, there is just one warning:
image

image

File: .github/workflows/ci.yml (hunk 6 above)
Comment (LucyJimenez) [reply .github/workflows/ci.yml:20]: Great, thanks for the explanation :)

File: .github/workflows/ci.yml (hunk 5 above)
Comment (LucyJimenez) [reply .github/workflows/ci.yml:11]: Thank you for pointing out that problem. Now the CI is working with Python 3.7 and 3.10. There is no warning reported on Python 3.7, and [two warnings](https://github.com/airspeed-velocity/asv/runs/5259920032?check_suite_focus=true) on Python 3.10. I'm going to check it in another PR.

image

image

**Comment by LucyJimenez** [comment]
Additional data that could be interesting about test:

- `CI Py2.7` 201 passed, 15 skipped in 466.88s (0:07:46)
- `CI Py3.10` 201 passed, 15 skipped, 2 warnings in 551.07s (0:09:11)
- `Appveyor` 168 passed, 48 skipped in 1259.22s (0:20:59)

**Comment by datapythonista** [comment]
Thanks @LucyJimenez, really nice

**Pull request merged**